test/demo/clear.inttest.ts

```diff
@@ -0,0 +1,254 @@
+import {testManagerIdentity, defaultTestAppUrl} from './utils/constants';
+import {delay, Duration} from './utils/delay';
+import {createApp} from './utils/spawnRemote';
+import * as notifsRemote from './utils/notificationsRemote';
+import {getAllToastWindows, getToastWindow} from './utils/toastUtils';
+import {Boxed} from './utils/types';
+import {setupWithCenterBookends, setupWithoutCenterBookends} from './common';
+
+const notificationWithoutOnCloseActionResult: NotificationOptions = {
+    body: 'Test Notification Body',
+    title: 'Test Notification Title',
+    category: 'Test Notification Category'
+};
+
+const notificationWithOnCloseActionResult1: NotificationOptions = {
+    body: 'Action Test Notification Body 1',
+    title: 'Action Test Notification Title 1',
+    category: 'Action Test Notification Category 1',
+    onClose: {task: 'close-1'}
+};
+
+const notificationWithOnCloseActionResult2: NotificationOptions = {
+    body: 'Action Test Notification Body 2',
+    title: 'Action Test Notification Title 2',
+    category: 'Action Test Notification Category 2',
+    onClose: {task: 'close-2'}
+};
+
+type SetupTestType = (notes: Notification[], indexToClear: number, testApp: Boxed<Application>) => void;
+
+type OuterTestParam = [
+    string,
+    () => void,
+    SetupTestType
+];
+
+type InnerTestParam = [
+    string,
+    NotificationOptions[],
+    number,
+    (CustomData | undefined)[]
+];
+
+const outerTestParams: OuterTestParam[] = [
+    [
+        'When clearing a notification with the Notification Center showing',
+        setupWithCenterBookends,
+        setupWithCenterClearedNotificationTest
+    ],
+    [
+        'When clearing a notification without the Notification Center showing',
+        setupWithoutCenterBookends,
+        setupWithoutCenterClearedNotificationTest
+    ]
+];
+
+const innerTestParams: InnerTestParam[] = [
+    [
+        'When clearing a notification, and no others have been created',
+        [notificationWithoutOnCloseActionResult],
+        0,
+        [undefined]
+    ],
+    [
+        'When clearing a notification with an `onClose` action result, and no other have been created',
+        [notificationWithOnCloseActionResult1],
+        0,
+        [{task: 'close-1'}]
+    ],
+    [
+        'When clearing a notification with an `onClose` action result, and one other notification has been created',
+        [notificationWithOnCloseActionResult1, notificationWithoutOnCloseActionResult],
+        0,
+        [{task: 'close-1'}, undefined]
+    ],
+    [
+        'When clearing a notification with an `onClose` action result, and one other notification with an `onClose` action result has been created',
+        [notificationWithOnCloseActionResult1, notificationWithOnCloseActionResult2],
+        1,
+        [undefined, {task: 'close-2'}]
+    ]
+];
+
+describe.each(outerTestParams)('%s', (titleParam: string, setupBookends: () => void, setupClearedNotificationTest: SetupTestType) => {
```

Comment:
Can we collapse a lot of this setup? A lot of additional types and scrolling involved, and the meaning of the items in the Inner/Outer TestParam types aren't that clear until you get down into the tests.

Have re-worked the outer loop, as an example:
```
descibe.each([
    ['When clearing all notifications with the Notification Center showing', true]},
    ['When clearing all notifications without the Notification Center showing', false]}
])('%s', (titleParam: string, centerOpen: boolean) => {
    if (centerOpen) {
        setupWithCenterBookends();

        test('The expected card has been removed from the Notification Center', async () => {
            await expect(getAllCenterCards()).resolves.toHaveLength(notes.length - 1);

            await expect(getCenterCardsByNotification(testApp.value.identity.uuid, notes[indexToClear].id)).resolves.toEqual([]);

            for (const note of notes.filter((note, index) => index !== indexToClear)) {
                await expect(getCenterCardsByNotification(testApp.value.identity.uuid, note.id)).resolves.toBeTruthy();
            }
        });
    } else {
        setupWithoutCenterBookends();

        test('The expected toast has been removed', async () => {
            await delay(Duration.TOAST_CLOSE);

            await expect(getAllToastWindows()).resolves.toHaveLength(notes.length - 1);

            await expect(getToastWindow(testApp.value.identity.uuid, notes[indexToClear].id)).resolves.toEqual(undefined);

            for (const note of notes.filter((note, index) => index !== indexToClear)) {
                await expect(getToastWindow(testApp.value.identity.uuid, note.id)).resolves.toBeTruthy();
            }
        });
    }

   // . . .
});
```
Not exactly short, but puts everything in one place - rather than types/data at the top, then the `each`, then helper functions at the bottom of the file. I _think_ this should also avoid the need to use boxing, as the 'setup' tests at the end of the file would then be defined in the same scope as the parameterised data.

Same pattern would work for `clearAll`, which also has some slightly-simpler setup from the look of it.

Reply:
I've improved the naming which should help things.

Hmmm, doing `if (...) {theseTests} else {thoseTests}` was what I was specifically trying to avoid. Talk when you've got a moment

Reply:
Changed as discussed

✅